### ISSUE TITLE
OS temperature

### ISSUE THREAD

**Issue by Amjad121092** [dblock/oshi#119]
Hi,

Would it be possible to add a method to get the PC temperature? I think this is crucial and it's missing. Other than that the project is amazing thank you very much!

Regards

**Comment by dblock** [comment]
I wanted to label this 🔥, but github doesn't support Emoji in issue labels!!!
https://twitter.com/dblockdotorg/status/701913056931135488

**Comment by dbwiddis** [comment]
This sounds like a great suggestion... question is how hard it is to get the info.

OS X looks the easiest.  [This project](https://github.com/lavoiesl/osx-cpu-temp) has C code that I can port to JNA.

Linux appears to require installation of a package lm-sensors, plus root permissions. 

Windows appears to have it (possibly) in WMI, info [here](http://stackoverflow.com/questions/23314886/get-cpu-temperature).

**Comment by dbwiddis** [comment]
Update... once lm-sensors is installed and sensors-detect is run as root, it does provide some basic info without root permissions.  So, it looks like this is a possibility!

**Comment by Amjad121092** [comment]
This is great news i hope you can add it in as it would be amazing to have!

**Comment by Amjad121092** [comment]
Hi,

Just wondering if there's any updates on this?

Regards

**Comment by dblock** [comment]
@Amjad121092 There's no update or ETA or anyone doing anything other than best effort to implement features. Please feel free to contribute, see https://github.com/dblock/oshi/blob/master/CONTRIBUTING.md.

**Comment by dbwiddis** [comment]
@Amjad121092 as @dblock said, this is a best effort approach.  I started on the OS X port and hit a snag (C pointers drive me nuts).  I haven't even looked at Linux and Windows yet.  The information is posted in this issue if you, or anyone you know, wants to take a crack at contributing a solution.

**Comment by dbwiddis** [comment]
Here's a rundown of current research/challenges.

Linux - This is looking the best so far.  If lm-sensors is installed by default (it is on Ubuntu) then the `sensors` command can be easily parsed.  By reverse engineering the lm-sensors code it appears that it gets the value from the path `/sys/class/hwmon/hwmon0/` and reading temp1_input.  There are other files in this directory structure that give information on fan speeds as well.  I haven't written any code here yet but it looks easy enough once the other OS's are taken care of.

OS X - I have ported code into JNA from the [iStats](https://github.com/Chris911/iStats) project.  In addition to querying the System Master Controller for temperatures, fan speed information is also available.   Despite using exactly the same code, my requests return a `kIOReturnIPCError` the overwhelming majority (99%) of the time.  I don't know why my code doesn't behave identically to the c code I'm porting from, or why it sometimes works.  I can ignore the IPC error and process the returned value anyway (there are other data checks to ensure I got readable results so this is totally reasonable); this seems to improve my chances.  I can regularly get results at least once if I retry about 30 times.   There may or may not be interference with another application I'm using to monitor my smc temperature and fan speeds.   In any case, I'm not sure we want to publish code which doesn't always work.   It would be more reliable to ask users to install the iStats code themselves and then just call `istats cpu temp` from the command line.  However, part of the original design of oshi was to avoid having to install anything.  I guess I'll keep banging my head against the wall of c vs. JNA to see why it's not working identically.

Windows - The information may be available in WMI and I've [located the appropriate code to query it](http://www.tomshardware.com/faq/id-1924879/learn-computer-command-prompt-wmic.html), however this is still just executing a command line.  Also, that site echoes what I've seen elsewhere, _"(note, doesn't work will all hardware or VMs)"_.  In particular, it doesn't work on the Windows VM I use for my own development/testing, nor my standalone windows laptop.  So even though I know what results to expect and how to parse them (when they exist) it's impossible for me to test, and falls into the same category of the OS X solution of only working some of the time.  Another option here is the [Open Hardware Monitor](http://openhardwaremonitor.org/) project.  This seems to work with a wider range of hardware and inserts its results into WMI where we could query it from OSHI.

So, in summary:
- Linux = possible using `sysfs` interface. 
- OS X = works sometimes.  Could document that users can install iStats to improve success.  Can keep trying to figure out why it's not working always.
- Windows = works on some hardware (but not mine so I can't test). Could document that users can install Open Hardware Monitor to expand hardware options.

Thoughts?

**Comment by dblock** [comment]
My thoughts: fun research! I am sure it's fixable, probably just needs a good night sleep and a good drink? :)

**Comment by dbwiddis** [comment]
I've tried the "sleep/drink" combo two consecutive nights.  Okay, maybe more drink/less sleep.  

I may or may not solve the OS X inconsistency (or maybe someone else can figure it out).

The bigger question applies specifically to Windows.  Ultimately this requires low-level hardware interaction (assembly language) that's impossible/impractical to replicate.  A very small percentage of Windows machines will provide this info in WMI by default.  Most will require installing the Open Hardware Monitor dll in order to get this info.   Oshi advertises itself as "no native dll required" and this will essentially require an asterisk for that statement.  Is that in line with the project vision?

**Comment by dblock** [comment]
@dbwiddis I think "no external native DLL required" is what we want for oshi, but I am 50/50 on this. You've been doing all the code, you should make the call.

**Comment by Amjad121092** [comment]
Hi,

I'll take a look into this later tonight, i'm unsure about the OS X issue but i might be able to solve the windows one. I'll come back with an update.

Regards

**Comment by dbwiddis** [comment]
@Amjad121092 thanks for looking into the windows portion.  My plan would be to first attempt to parse these command lines:
`wmic /namespace:\\root\wmi PATH MSAcpi_ThermalZoneTemperature get CurrentTemperature /value'
if that failed:
`wmic /namespace:\\root\wmi PATH Win32_TemperatureProbe get CurrentReading /value`
or possibly:
`wmic temperature get CurrentReading`

Looking further at Open Hardware Monitor, it's not a dll... the .exe program would actually have to be running (potentially set up as a service) to generate the WMI that oshi could use if all the above failed.   I think it's reasonable to simply say "this rarely works on Windows but if you run this program, oshi will parse it."  So it's not a requirement, it's simply an option to enable an otherwise inaccessible value.

**Comment by dblock** [comment]
You might want to look into the many attempts at querying WMI info from Java, like http://henryranch.net/software/jwmi-query-windows-wmi-from-java (I just googled it). It seems totally possible to do this in JNA proper (there has been a lot of COM support added recently) and would be a very valuable contribution to the JNA project.

**Comment by dbwiddis** [comment]
I looked at jWMI before... under the hood it just calls a command line executable anyway, so no better than executing wmic.exe (except expands to older systems.)  It's not so much querying WMI that's the issue, it's the fact that very few processor vendors actually publish their stats to WMI.

There are several temperature monitoring programs out there and it's not too hard to figure out what commands need to be sent for the major processors (Intel, AMD).  It's a "juice worth the squeeze" argument.

In other news, I've gotten the OS X code working more reliably by adding delays strategically.  Still can't guarantee it works on the first try like the C code, which is frustrating.  I wonder why going via JNA would introduce communication issues.

**Comment by dbwiddis** [comment]
Mac and Linux code in #121 for review while we figure out what to do with Windows.

**Comment by dbwiddis** [comment]
Status update:   Updated OS X and Linux code to get CPU Temperature, fan speed (for all fans), and CPU input voltage.   OS X code is touchy with the inter process communication, but I've tweaked it with some processing delays it works most of the time.  Linux works great (if info is available).

Windows is the toughest nut to crack.  I've read pages and pages of search results and come to the conclusion that there are two ways of doing it:
* Using WMI.  This works rarely. MS has provided the API but motherboard manufacturers aren't using it.  There seem to be two options for this, the root/wmi namespace (older) and the root/cimv2 namespace (newer).  
* Directly querying the processor with a `cpuid` opcode.  Which code to use, and what math to use with it to calculate temperature, varies based on which processor is used.  

The [Open Hardware Monitor](http://openhardwaremonitor.org/) project has done the heavy lifting to figure out what to do (see, for example, [here](https://github.com/openhardwaremonitor/openhardwaremonitor/blob/master/Hardware/CPU/IntelCPU.cs), starting at line 82).  This gives an idea of the scale of the problem.

We can attempt existing WMI and:
* Throw up our hands if we can't figure it out, just making this a data field Windows doesn't provide.
* Test the WMI output by Open Hardware Monitor.  If it's running, we can use those values in oshi.  (Requires user to be running OHM.)  Otherwise we're no worse off than before.
* Attempt to back-port the OHM code into Java using JNA.  I don't even know where to start using processor opcodes; plus we'd have to meticulously copy-and-update all the OHM code (see above).  I don't have the time or interest to do that. :)

**Comment by dbwiddis** [comment]
I went ahead and implemented command-line-parsing for wmi so far.   Leaving this issue open to add Open Hardware Monitor WMI information if available, and possibly look at other ways to access wmi than the command line.

**Comment by dbwiddis** [comment]
Closing this issue.  I briefly tried using the COM interface to get WMI programmatically but it's a bridge too far for me.  Maybe someone else can do it someday.
